src/event_manager/subscribers.rs

```diff
@@ -0,0 +1,51 @@
+use super::SubscriberId;
+use std::collections::HashMap;
+
+// Internal structure used to keep the set of subscribers registered with an EventManger.
+// This structure is a thin wrapper over a `HashMap` in which the keys are uniquely
+// generated when calling `add`.
+pub(crate) struct Subscribers<T> {
+    // The key is the unique id of the subscriber and the entry is the `Subscriber`.
+    subscribers: HashMap<SubscriberId, T>,
+    // We are generating the unique ids by incrementing this counter for each added subscriber,
+    // and rely on the extremely large value range of u128 to ensure each value is effectively
+    // unique over the runtime of any VMM.
+    next_id: u128,
+}
+
+impl<T> Subscribers<T> {
+    pub(crate) fn new() -> Self {
+        Subscribers {
+            subscribers: HashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    // Adds a subscriber and generates an unique id to represent it.
+    pub(crate) fn add(&mut self, subscriber: T) -> SubscriberId {
+        let id = SubscriberId(self.next_id);
+        self.next_id += 1;
+
+        self.subscribers.insert(id, subscriber);
+
+        id
+    }
+
+    // Remove and return the subscriber associated with the given id, if it exists.
+    pub(crate) fn remove(&mut self, subscriber_id: SubscriberId) -> Option<T> {
+        self.subscribers.remove(&subscriber_id)
+    }
+
+    // Checks whether a subscriber with `subscriber_id` is registered.
+    pub(crate) fn contains(&mut self, subscriber_id: SubscriberId) -> bool {
+        self.subscribers.contains_key(&subscriber_id)
+    }
+
+    // Return a mutable reference to the subriber represented by `subscriber_id`.
+    //
+    // This method should only be called for indices that are known to be valid, otherwise
+    // panics can occur.
+    pub(crate) fn get_mut(&mut self, subscriber_id: SubscriberId) -> &mut T {
```

Comment:
Since this does `unwrap` without any checks, I suggest we change its name to `fn get_mut_unchecked()`.

```suggestion
    pub(crate) fn get_mut_unchecked(&mut self, subscriber_id: SubscriberId) -> &mut T {
```